File: sdk/python/coreauth/services/oauth2.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional
from urllib.parse import urlencode

if TYPE_CHECKING:
    from .._http import HttpClient


class OAuth2Service:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def discovery(self) -> dict:
        return self._http.get("/.well-known/openid-configuration")

    def jwks(self) -> dict:
        return self._http.get("/.well-known/jwks.json")

    def authorize(
        self,
        client_id: str,
        redirect_uri: str,
        response_type: str = "code",
        scope: str = "openid",
        **params: Any,
    ) -> str:
        query: dict[str, Any] = {
            "client_id": client_id,
            "redirect_uri": redirect_uri,
            "response_type": response_type,
            "scope": scope,
            **params,
        }
        base_url = self._http._base_url.rstrip("/")
        return f"{base_url}/authorize?{urlencode(query)}"

    def token(self, grant_type: str, **kwargs: Any) -> dict:
        data = {"grant_type": grant_type, **kwargs}
        return self._http.post_form("/oauth/token", data=data)

    def userinfo(self) -> dict:
        return self._http.get("/userinfo")

    def revoke(self, token: str, token_type_hint: Optional[str] = None) -> dict:
        data: dict[str, str] = {"token": token}
        if token_type_hint:
            data["token_type_hint"] = token_type_hint
        return self._http.post_form("/oauth/revoke", data=data)

    def introspect(self, token: str, token_type_hint: Optional[str] = None) -> dict:
        data: dict[str, str] = {"token": token}
        if token_type_hint:
            data["token_type_hint"] = token_type_hint
        return self._http.post_form("/oauth/introspect", data=data)

    def oidc_logout(
        self,
        id_token_hint: Optional[str] = None,
        post_logout_redirect_uri: Optional[str] = None,
        state: Optional[str] = None,
    ) -> dict:
        params: dict[str, str] = {}
        if id_token_hint:
            params["id_token_hint"] = id_token_hint
        if post_logout_redirect_uri:
            params["post_logout_redirect_uri"] = post_logout_redirect_uri
        if state:
            params["state"] = state
        return self._http.get("/logout", params=params or None)

File: sdk/node/src/services/oauth2.ts
import { HttpClient } from '../http.js';

export class OAuth2Service {
  constructor(private http: HttpClient) {}

  discovery(): Promise<any> {
    return this.http.get('/.well-known/openid-configuration');
  }

  jwks(): Promise<any> {
    return this.http.get('/.well-known/jwks.json');
  }

  authorize(
    clientId: string,
    redirectUri: string,
    responseType: string = 'code',
    scope: string = 'openid',
    params?: Record<string, string>,
  ): string {
    const query: Record<string, string> = {
      client_id: clientId,
      redirect_uri: redirectUri,
      response_type: responseType,
      scope,
      ...params,
    };
    const baseUrl = ((this.http as any).baseUrl as string).replace(/\/+$/, '');
    return `${baseUrl}/authorize?${new URLSearchParams(query).toString()}`;
  }

  token(grantType: string, params?: Record<string, string>): Promise<any> {
    const data: Record<string, string> = { grant_type: grantType, ...params };
    return this.http.postForm('/oauth/token', data);
  }

  userinfo(): Promise<any> {
    return this.http.get('/userinfo');
  }

  revoke(token: string, tokenTypeHint?: string): Promise<any> {
    const data: Record<string, string> = { token };
    if (tokenTypeHint) data.token_type_hint = tokenTypeHint;
    return this.http.postForm('/oauth/revoke', data);
  }

  introspect(token: string, tokenTypeHint?: string): Promise<any> {
    const data: Record<string, string> = { token };
    if (tokenTypeHint) data.token_type_hint = tokenTypeHint;
    return this.http.postForm('/oauth/introspect', data);
  }

  oidcLogout(params?: {
    id_token_hint?: string;
    post_logout_redirect_uri?: string;
    state?: string;
  }): Promise<any> {
    return this.http.get('/logout', params);
  }
}
